Errors in build.py not logged helpfully in "BUILD FAILED" error message

Right now, when there is an error in the `build.py` script itself, the "BUILD FAILED" error message does not provide the full stack trace, nor does the logger provide information about the *type* of exception or line number. For example:

### An `AttributeError` example in `build.py`

Given a sample task of:
```python
@task('analyze')
def sample_task():
    import os
    os.getcd()
```

Where `os.getcwd()` is replaced with the incorrect `os.getcd()`, running `pyb` will print the following error message:
> BUILD FAILED - module 'os' has no attribute 'getcd'

A more helpful error message would be:

> BUILD FAILED - AttributeError: module 'os' has no attribute 'getcd' (build.py:37)

Where `AttributeError` is the *type* of exception, followed by the message, followed by the line number.

### An `KeyError` example in `build.py`

Given a task of:
```python
@task('analyze')
def sample_task():
    d = {'foo': 1, 'bar': 2}
    print(d['baz'])
```
The failure message will be:

> BUILD FAILED - 'baz'

A more helpful error message would be:

> BUILD FAILED - KeyError: 'baz' (build.py:22)

Where `KeyError` is the type of exception and 22 is the line number.

## Potential Solution 

Using `sys.exc_info()` to capture a tuple containing the exception type and traceback can provide the needed functionality to implement this. A sample implementation in the [`main` method of `clip.py`](https://github.com/pybuilder/pybuilder/blob/772cf66a6fea86c59bd76f22388b0ce964b2fc1a/src/main/python/pybuilder/cli.py#L385) might look something like:

```python
        except PyBuilderException as e:
            exc_type, exc_obj, exc_tb = sys.exc_info()

            filename = exc_tb.tb_frame.f_code.co_filename
            lineno = exc_tb.tb_lineno

            msg = "{exc_type}: {e} ({filename}:{lineno})".format(
                exc_type=exc_type, e=e, filename=filename, lineno=lineno
            )

            print_build_status(str(msg), options, successful=False)
            return 1
```

--- a/src/main/python/pybuilder/cli.py
+++ b/src/main/python/pybuilder/cli.py
@@ -26,6 +26,7 @@ import optparse
 import re
 import sys
 import traceback
+from os.path import sep, normcase as nc
 
 from pybuilder import __version__
 from pybuilder import extern
@@ -371,6 +372,33 @@ def print_plan_list_of_tasks(options, arguments, reactor, quiet=False):
     print_task_list(execution_plan, quiet)
 
 
+def get_failure_message():
+    exc_type, exc_obj, exc_tb = sys.exc_info()
+
+    filename = None
+    lineno = None
+
+    while exc_tb.tb_next:
+        exc_tb = exc_tb.tb_next
+
+    frame = exc_tb.tb_frame
+    if hasattr(frame, "f_code"):
+        code = frame.f_code
+        filename = code.co_filename
+        lineno = exc_tb.tb_lineno
+
+        filename = nc(filename)
+        for path in sys.path:
+            path = nc(path)
+            if filename.startswith(path) and len(filename) > len(path) and filename[len(path)] == sep:
+                filename = filename[len(path) + 1:]
+                break
+
+    return "%s%s%s" % ("%s: " % exc_type.__name__ if not isinstance(exc_obj, PyBuilderException) else "",
+                       exc_obj,
+                       " (%s:%d)" % (filename, lineno) if filename else "")
+
+
 def main(*args):
     if not args:
         args = sys.argv[1:]
@@ -407,8 +435,8 @@ def main(*args):
             if options.list_plan_tasks:
                 print_plan_list_of_tasks(options, arguments, reactor, quiet=options.very_quiet)
             return 0
-        except PyBuilderException as e:
-            print_build_status(str(e), options, successful=False)
+        except PyBuilderException:
+            print_build_status(get_failure_message(), options, successful=False)
             return 1
 
     if not options.very_quiet:
@@ -442,9 +470,9 @@ def main(*args):
         except KeyboardInterrupt:
             raise PyBuilderException("Build aborted")
 
-    except (Exception, SystemExit) as e:
+    except (Exception, SystemExit):
         successful = False
-        failure_message = str(e)
+        failure_message = get_failure_message()
         if options.debug:
             traceback.print_exc(file=sys.stderr)
 

--- a/src/unittest/python/cli_tests.py
+++ b/src/unittest/python/cli_tests.py
@@ -23,8 +23,10 @@ from pybuilder.cli import (parse_options,
                            CommandLineUsageException,
                            StdOutLogger,
                            length_of_longest_string,
-                           print_list_of_tasks)
+                           print_list_of_tasks,
+                           get_failure_message)
 from pybuilder.core import Logger
+from pybuilder.errors import PyBuilderException
 from test_utils import Mock, patch, call
 
 
@@ -232,3 +234,17 @@ class LengthOfLongestStringTests(unittest.TestCase):
 
     def test_should_return_four_when_list_contains_foo_bar_egg_and_spam(self):
         self.assertEqual(4, length_of_longest_string(["egg", "spam", "foo", "bar"]))
+
+
+class ErrorHandlingTests(unittest.TestCase):
+    def test_generic_error_message(self):
+        try:
+            raise Exception("test")
+        except Exception:
+            self.assertRegexpMatches(get_failure_message(), r"Exception: test \(cli_tests.py\:\d+\)")
+
+    def test_pyb_error_message(self):
+        try:
+            raise PyBuilderException("test")
+        except Exception:
+            self.assertRegexpMatches(get_failure_message(), r"test \(cli_tests.py\:\d+\)")

